### PR TITLE
Add support for selective loading of broadcast datasources in the task layer

### PR DESCRIPTION
**Pull request by abhishekrb19**
### Motivation

Currently, servers and tasks download all broadcast datasources. While the downloading of bootstrap segments is generally fast, as it happens in parallel during startup, it does consume storage space. This is especially relevant for tasks like kill tasks and MSQ worker tasks, which either don't need to load any broadcast datasources or only selectively load those that are required. This optimization will save task storage space and speed up the task startup time a bit.

### Description

The design for loading broadcast datasources follows a similar approach to what was implemented for loading lookups in https://github.com/apache/druid/pull/16328. 

Broadcast datasources can be specified in SQL queries through `JOIN` and `FROM` clauses, or obtained from other sources such as lookups.To this effect, we have introduced a `BroadcastDatasourceLoadingSpec`. Finding the set of broadcast datasources during SQL planning will be done in a follow-up, which will apply only to MSQ tasks,  so they load only required broadcast datasources. This PR primarily focuses on the skeletal changes around `BroadcastDatasourceLoadingSpec` and integrating it from the `Task` interface via `CliPeon` to `SegmentBootstrapper`.

#### Changes

- Added `BroadcastDatasourceLoadingSpec` class and `Task.getBroadcastDatasourceLoadingSpec()` tasks to control how broadcast datasources are loaded.
   - The supported broadcast datasource loading modes are `ALL`, `NONE` and `ONLY_REQUIRED`.
   - By default, all tasks will load all broadcast datasources. 
   - The `KillTask` and `MSQControllerTask` override `getBroadcastDatasourceLoadingSpec()` to `NONE`.
- The `CliPeon` initializes the broadcast module by parsing the command line option `--loadBroadcastDatasourceMode`.
- The CLIPeon option `--loadBroadcastSegments` have been marked as deprecated in favor of `--loadBroadcastDatasourceMode`. This can be removed in a future release.
- The `PodTemplateTaskAdapter` and `K8sTaskAdapter` support wiring for the option `LOAD_BROADCAST_DATASOURCE_MODE`, similar to existing `LOAD_BROADCAST_SEGMENTS`.

In summary:
- Data servers load _all_ broadcast datasources.
- By default, all tasks load _all_ broadcast datasources unless an override exists.
- MSQ Controller and Kill tasks do _not_ load any broadcast datasources.
- In a future patch, MSQ ingest and query tasks will _selectively_ load segments pertaining to the broadcast datasources determined as part of query planning. This will be addressed in a follow-up.

#### Release note:

Tasks control the loading of broadcast datasources via `BroadcastDatasourceLoadingSpec getBroadcastDatasourceLoadingSpec()`. By default, tasks download all broadcast datasources, unless there's an override as with kill and MSQ controller task. 

The `CLIPeon` command line option `--loadBroadcastSegments` is deprecated in favor of `--loadBroadcastDatasourceMode`.

This PR has:

- [x] been self-reviewed.
- [x] added documentation for new or modified features or behaviors.
- [x] a release note entry in the PR description.
- [x] added Javadocs for most classes and all non-trivial methods. Linked related entities via Javadoc links.
- [x] added comments explaining the "why" and the intent of the code wherever would not be obvious for an unfamiliar reader.
- [x] added unit tests or modified existing tests to cover new code paths, ensuring the threshold for [code coverage](https://github.com/apache/druid/blob/master/dev/code-review/code-coverage.md) is met.
- [x] been tested in a test Druid cluster.

### CODE REVIEW

File: extensions-contrib/kubernetes-overlord-extensions/src/main/java/org/apache/druid/k8s/overlord/taskadapter/K8sTaskAdapter.java [hunk 1]
```diff
@@ -444,12 +444,16 @@ private List<String> generateCommand(Task task)
     }
 
     // If the task type is queryable, we need to load broadcast segments on the peon, used for
-    // join queries
+    // join queries. This is replaced by --loadBroadcastDatasourceMode option, but is preserved here
+    // for backwards compatibility and can be removed in a future release.
     if (task.supportsQueries()) {
       command.add("--loadBroadcastSegments");
```
Comment (georgew5656) [inline extensions-contrib/kubernetes-overlord-extensions/src/main/java/org/apache/druid/k8s/overlord/taskadapter/K8sTaskAdapter.java:450]: i think if this defaults to true on peons we can just remove this

Comment (georgew5656) [reply extensions-contrib/kubernetes-overlord-extensions/src/main/java/org/apache/druid/k8s/overlord/taskadapter/K8sTaskAdapter.java:450]: looks like it defaults to false, probably best to leave it in then.

**Review by georgew5656 — APPROVED** [review]

File: indexing-service/src/main/java/org/apache/druid/indexing/common/task/Task.java [hunk 2]
```diff
@@ -184,6 +186,7 @@ default Set<ResourceAction> getInputSourceResources() throws UOE
    * MSQWorkerTask returns true from this method (because it embeds a query stack for running multi-stage queries)
    * even though it is not directly queryable via HTTP.
    */
+  @Deprecated
```
Comment (cryptoe) [inline indexing-service/src/main/java/org/apache/druid/indexing/common/task/Task.java:189]: We adjust merge pools and stuff with this method. Why is this marked deprecated. The javadocs below fail to shed light on this.

File: server/src/main/java/org/apache/druid/server/coordination/SegmentBootstrapper.java [hunk 3]
```diff
@@ -261,9 +267,16 @@ private void loadSegmentsOnStartup() throws IOException
 
   /**
    * @return a list of bootstrap segments. When bootstrap segments cannot be found, an empty list is returned.
+   * The bootstrap segments returned are filtered by the broadcast datasources indicated by {@link DataSourceTaskIdHolder#getBroadcastDatasourceLoadingSpec()}
+   * if applicable.
    */
   private List<DataSegment> getBootstrapSegments()
   {
+    final BroadcastDatasourceLoadingSpec.Mode mode = datasourceHolder.getBroadcastDatasourceLoadingSpec().getMode();
```
Comment (cryptoe) [inline server/src/main/java/org/apache/druid/server/coordination/SegmentBootstrapper.java:275]: Can we use datasourceHolder.getLookupLoadingSpec() here and not add any new configs  ?

File: indexing-service/src/main/java/org/apache/druid/indexing/common/task/Task.java (hunk 2 above)
Comment (abhishekrb19) [reply indexing-service/src/main/java/org/apache/druid/indexing/common/task/Task.java]: Ah that's right, I see it in the usages. Removed the `@Deprecated` annotation.

File: server/src/main/java/org/apache/druid/server/coordination/SegmentBootstrapper.java (hunk 3 above)
Comment (abhishekrb19) [reply server/src/main/java/org/apache/druid/server/coordination/SegmentBootstrapper.java:275]: related comment: https://github.com/apache/druid/pull/17027#discussion_r1753165605

File: extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/indexing/MSQControllerTask.java [hunk 4]
```diff
@@ -374,4 +375,10 @@ public LookupLoadingSpec getLookupLoadingSpec()
   {
     return LookupLoadingSpec.NONE;
   }
+
+  @Override
+  public BroadcastDatasourceLoadingSpec getBroadcastDatasourceLoadingSpec()
+  {
+    return BroadcastDatasourceLoadingSpec.NONE;
```
[the remark this answers is not in the document]
Comment (cryptoe) [reply extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/indexing/MSQControllerTask.java:382]: I mean can we rename the LookingLoadingSpec to me more generic and then reuse the class in both places no?

[the remark this answers is not in the document]
Comment (abhishekrb19) [reply extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/indexing/MSQControllerTask.java:382]: yeah, thinking about this, the `LookingLoadingSpec` is used in the [Task](https://github.com/apache/druid/blob/master/indexing-service/src/main/java/org/apache/druid/indexing/common/task/Task.java#L333) interface, so it's not trivial to rename and merge the functionalities. Although, I think there's some scope to extract some common things like `Mode` to a static class that can be shared between `LookingLoadingSpec` and `BroadcastDatasourceLoadingSpec`. I can handle that in the next PR.

**Comment by abhishekrb19** [comment]
The failing UTs are unrelated: other module test failed with OOM (ongoing investigation with heap dumps from https://github.com/apache/druid/pull/17029) and flaky CDS test.

**Pull request merged**